lib/phoenix_test/playwright/cookies.ex

```diff
@@ -0,0 +1,93 @@
+
+  ## Cookie fields
+
+  | key          | type        | description |
+  | -----------  | ----------- | ----------- |
+  | `:name`      | `binary()`  | |
+  | `:value`     | `binary()`  | |
+  | `:url`       | `binary()`  | *(optional)* either url or domain / path are required |
+  | `:domain`    | `binary()`  | *(optional)* either url or domain / path are required |
+  | `:path`      | `binary()`  | *(optional)* either url or domain / path are required |
+  | `:expires`   | `float()`   | *(optional)* Unix time in seconds. |
+  | `:http_only` | `boolean()` | *(optional)* |
+  | `:secure`    | `boolean()` | *(optional)* |
+  | `:same_site` | `binary()`  | *(optional)* one of "Strict", "Lax", "None" |
+  """
+
+  alias Plug.Conn
+  alias Plug.Session
+
+  @type cookie :: %{
+          :name => binary(),
+          :value => binary() | map(),
+          :url => binary(),
+          :domain => binary(),
+          :path => binary(),
+          :expires => float(),
+          :http_only => boolean(),
+          :secure => boolean(),
+          :same_site => binary()
+        }
+
+  @doc """
+  Converts the cookie map into a string-keyed map suitable for posting
+  """
+  def to_params_map(cookie) do
+    Map.update(cookie, :value, "", fn value ->
+      otp_app = Application.get_env(:phoenix_test, :otp_app)
+      endpoint = Application.get_env(:phoenix_test, :endpoint)
+      secret_key_base = Application.get_env(otp_app, endpoint)[:secret_key_base]
+
+      opts =
+        cookie
+        |> Map.take([:domain, :max_age, :path, :http_only, :secure, :extra, :sign, :encrypt, :same_site])
+        |> Map.to_list()
+
+      plug_cookie =
+        Conn.put_resp_cookie(%Conn{secret_key_base: secret_key_base}, to_string(cookie.name), value, opts)
+
+      plug_cookie.resp_cookies[cookie.name].value
+    end)
+  end
+
+  @doc """
+  Converts the session cookie map (with value that is also a map) into a string-keyed map suitable for posting
+  """
+  def to_session_params_map(cookie, session_options) do
+    cookie
+    |> Map.update(:value, "", fn value ->
+      otp_app = Application.get_env(:phoenix_test, :otp_app)
+      endpoint = Application.get_env(:phoenix_test, :endpoint)
+      secret_key_base = Application.get_env(otp_app, endpoint)[:secret_key_base]
+
+      %Conn{secret_key_base: secret_key_base, owner: self()}
+      |> Session.call(Session.init(session_options))
+      |> Conn.fetch_session()
+      |> then(fn conn ->
+        Enum.reduce(value, conn, fn {key, val}, conn ->
+          Conn.put_session(conn, key, val)
+        end)
+      end)
+      |> Conn.fetch_cookies(signed: [session_options[:key]])
+      |> Map.update!(:adapter, fn {_adapter, nil} ->
+        {PhoenixTest.Playwright.Cookies.PseudoAdapter, nil}
+      end)
+      |> Conn.send_resp(200, "")
+      |> Map.get(:cookies)
+      |> Map.get(session_options[:key])
+    end)
+    |> Map.update(:name, session_options[:key], & &1)
```

Comment:
Great to see you got it working. Tough one. Need to wrap my head around it.
And I'm sure there's some readability gains to be had by refactoring in a next step.

Reply:
I will give it a gentle refactor when I'm addressing the other comments. Probably won't be until Friday because I have a lot on tomorrow.